docs/dns-policy.md

```diff
@@ -0,0 +1,497 @@
+This policy is used to provide dns management for gateway listeners by managing the lifecycle of dns records in external dns providers such as AWS Route53 and Google DNS.
+
+## Terms
+
+- [`GatewayAPI`](https://gateway-api.sigs.k8s.io/): resources that model service networking in Kubernetes.
+- [`Gateway`](https://gateway-api.sigs.k8s.io/api-types/gateway/): Kubernetes Gateway resource. 
+- [`ManagedZone`](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/config/crd/bases/kuadrant.io_managedzones.yaml): Kuadrant resource representing an Apex Zone in a dns provider.
+- [`DNSPolicy`](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/config/crd/bases/kuadrant.io_dnspolicies.yaml): Kuadrant policy for managing gateway dns.
+- [`DNSRecord`](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/config/crd/bases/kuadrant.io_dnsrecords.yaml): Kuadrant resource representing a set of records in a managed zone.
+
+## DNS Provider Setup
+
+A DNSPolicy acts against a target Gateway by processing its listeners for hostnames that it can create dns records for. In order for it to do this, it must know about dns providers, and what domains these dns providers are currently hosting.
+This is done through the creation of ManagedZones and dns provider secrets containing the credentials for the dns provider account.
+
+If for example a Gateway is created with a listener with a hostname of `echo.apps.hcpapps.net`:
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+  namespace: multi-cluster-gateways
+spec:
+  gatewayClassName: kuadrant-multi-cluster-gateway-instance-per-cluster
+  listeners:
+    - allowedRoutes:
+        namespaces:
+          from: All
+      name: api
+      hostname: echo.apps.hcpapps.net
+      port: 443
+      protocol: HTTPS
+```
+
+In order for the DNSPolicy to act upon that listener, a ManagedZone must exist for that hostnames domain. 
+
+A secret containing the provider credentials must first be created:
+```bash
+kubectl create secret generic my-aws-credentials --type=kuadrant.io/aws --from-env-file=./aws-credentials.env -n multi-cluster-gateways
+kubectl get secrets my-aws-credentials -n multi-cluster-gateways -o yaml
+apiVersion: v1
+data:
+  AWS_ACCESS_KEY_ID: <AWS_ACCESS_KEY_ID>
+  AWS_REGION: <AWS_REGION>
+  AWS_SECRET_ACCESS_KEY: <AWS_SECRET_ACCESS_KEY>
+kind: Secret
+metadata:
+  name: my-aws-credentials
+  namespace: multi-cluster-gateways
+type: kuadrant.io/aw
+```
+
+And then a ManagedZone can be added for the desired domain referencing the provider credentials:
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: ManagedZone
+metadata:
+  name: apps.hcpapps.net
+  namespace: multi-cluster-gateways
+spec:
+  domainName: apps.hcpapps.net
+  description: "apps.hcpapps.net managed domain"
+  dnsProviderSecretRef:
+    name: my-aws-credentials
+    namespace: multi-cluster-gateways
+    type: AWS
```

Comment:
A comment on the ManagedZone recourse here, not the example.
Is the `type` field the type of the Secret or the type of the dns provider?
Maybe it's both and doesn't matter?

Reply:
Good question, i don't believe this is actually required, and probably should just be removed from the spec. /cc @R-Lawton

Reply:
Yeah not sure this is needed. But it is a separate issue. Initially we had a different DNSProvider API that specified the provider but we have since moved to a secret with the type in it, so I don't think we need the type in the MZ. In additon we may shortly drop the MZ altogether see https://github.com/orgs/Kuadrant/discussions/7

Reply:
Its the type of dnsprovider. It was added to make it easier for users to see what provider is being used without having to go digging but it can be removed